lib/spack/spack/database.py

```diff
@@ -209,7 +209,12 @@ def to_dict(self, include_fields=default_install_record_fields):
 
         for field_name in include_fields:
             if field_name == 'spec':
-                rec_dict.update({'spec': self.spec.to_node_dict()})
+                spec_dict_list = self.spec.to_dict()['spec']
+                spec_dict = {}
+                for d in spec_dict_list:
```

Comment:
This [comment](https://github.com/spack/spack/blob/89332361ba2a1c55ee3d0702ae517b10964535c9/lib/spack/spack/spec.py#L1621) made me wonder if we could skip iterating down the list returned by `to_dict()` here, and instead just access the first item?

Reply:
The database is used to write both the traditional database index, as well as (more recently) the buildcache index.  The database `_write_to_file` method uses `InstallRecord.to_dict` on the specs it writes to the file.  So that part is needed and working for my purposes (of getting the `full_hash` written to both files).

> e.g. `InstallRecord.from_dict` discards the stored hash in favor of the one provided as a parameter to the function

I wanted to avoid changing how the database works for tracking installed specs, mostly because I didn't want to break anything.  But for reading back in a buildcache index, I'm using the database class (see [here](https://github.com/spack/spack/pull/19209/files#diff-d70c217a1f2343039a9033d745cf5a0517ac0fe42857dd8a5be6208bd4dc5b63R115-R123) for example), and it must get me the `full_hash` stored in the index rather than compute it anew.  Do we want similar functionality for tracking installed specs?  If so, then I should probably make sure the stored `full_hash` makes it from the index file back into the live `InstallRecord` instances.